File: web/src/lib/streakRepository.ts
import {
  doc,
  onSnapshot,
  runTransaction,
  serverTimestamp,
  Timestamp,
  type Unsubscribe,
} from "firebase/firestore";
import { db } from "./firebaseConfig";
import type { StreakSummary } from "../types/streak";

function streakDoc(groupId: string) {
  return doc(db, "syncGroups", groupId, "streaks", "summary");
}

/**
 * Subscribes to live updates of the group's streak summary. The doc may not
 * exist yet (no task list has ever been fully cleared) -- callback receives
 * the zero-value default in that case rather than null, so callers don't
 * each need to repeat that fallback.
 */
export function subscribeToStreak(
  groupId: string,
  callback: (summary: StreakSummary) => void,
): Unsubscribe {
  return onSnapshot(streakDoc(groupId), (snap) => {
    const data = snap.data() as Partial<StreakSummary> | undefined;
    callback({
      currentStreak: data?.currentStreak ?? 0,
      lastCompletedAllAt: (data?.lastCompletedAllAt as Timestamp | null | undefined) ?? null,
    });
  });
}

/** Local (browser-timezone) calendar date, stripped of time-of-day. */
function localCalendarDate(date: Date): Date {
  return new Date(date.getFullYear(), date.getMonth(), date.getDate());
}

function sameCalendarDate(a: Date, b: Date): boolean {
  return (
    a.getFullYear() === b.getFullYear() &&
    a.getMonth() === b.getMonth() &&
    a.getDate() === b.getDate()
  );
}

/**
 * Records that every task in the group was just marked complete, updating the
 * daily streak. Meant to be called exactly once per "list just cleared"
 * transition (incomplete count going from >0 to 0) -- see useTasks.ts.
 *
 * Uses the LOCAL calendar day deliberately (not UTC): this is a personal
 * daily-habit streak for a single user/device pair, not something that needs
 * cross-timezone consistency.
 *
 * Runs in a Firestore transaction so concurrent completions from paired
 * devices can't double-increment the streak. Returns the resulting streak
 * count (unchanged if today was already recorded).
 */
export async function recordAllTasksCleared(groupId: string): Promise<number> {
  const today = localCalendarDate(new Date());
  const ref = streakDoc(groupId);

  return runTransaction(db, async (transaction) => {
    const snap = await transaction.get(ref);
    const data = snap.data() as Partial<StreakSummary> | undefined;

    const currentStreak = data?.currentStreak ?? 0;
    const lastCompletedAllAt = (data?.lastCompletedAllAt as Timestamp | null | undefined) ?? null;
    const lastDate = lastCompletedAllAt ? localCalendarDate(lastCompletedAllAt.toDate()) : null;

    if (lastDate && sameCalendarDate(lastDate, today)) {
      // Already recorded today -- avoid double-counting if the user
      // un-completes and re-completes the last task the same day.
      return currentStreak;
    }

    const yesterday = new Date(today);
    yesterday.setDate(yesterday.getDate() - 1);

    const newStreak =
      lastDate && sameCalendarDate(lastDate, yesterday) ? currentStreak + 1 : 1;

    transaction.set(
      ref,
      { currentStreak: newStreak, lastCompletedAllAt: serverTimestamp() },
      { merge: true },
    );

    return newStreak;
  });
}

File: android/app/src/main/java/com/wikzo/todo/data/repository/StreakRepository.kt
package com.wikzo.todo.data.repository

import com.google.firebase.Timestamp
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions
import kotlinx.coroutines.tasks.await
import java.time.LocalDate
import java.time.ZoneId
import javax.inject.Inject
import javax.inject.Singleton

private const val FIELD_CURRENT_STREAK = "currentStreak"
private const val FIELD_LAST_COMPLETED_ALL_AT = "lastCompletedAllAt"

/**
 * Maintains `syncGroups/{groupId}/streaks/summary`, per /docs/data-model.md.
 *
 * The only mutation this repository performs is [recordAllTasksCleared], called
 * once per "the incomplete-task count just dropped from >0 to 0" moment -- the same
 * event that triggers the confetti celebration in TaskListViewModel. Everything
 * else about a task's lifecycle (add/edit/complete/delete) stays in [TaskRepository];
 * this is a separate, narrower doc so the two concerns don't get tangled in one
 * class.
 */
@Singleton
class StreakRepository @Inject constructor(
    private val firestore: FirebaseFirestore,
) {

    private fun summaryDoc(groupId: String) =
        firestore.collection("syncGroups").document(groupId)
            .collection("streaks").document("summary")

    /**
     * Records that every task in [groupId] just became complete, and returns the
     * resulting streak count.
     *
     * Runs as a Firestore transaction (read-then-conditionally-write) so two
     * near-simultaneous clears -- e.g. this device and a paired device racing to
     * complete the last task -- can't corrupt the count. Algorithm, using the
     * *device's local calendar day* (deliberately not UTC -- this is a personal
     * daily-habit streak, not something needing cross-timezone consistency):
     *
     * 1. If the doc's `lastCompletedAllAt` is already today, this is a no-op --
     *    the user un-completed and re-completed the last task the same day, and
     *    the existing streak is returned unchanged rather than bumped again.
     * 2. Otherwise, if `lastCompletedAllAt` was yesterday, the streak continues
     *    (`currentStreak + 1`); any other gap (or no prior doc) resets it to 1.
     * 3. The doc is written with `SetOptions.merge()`, not `set()`/`update()`, so
     *    it doesn't clobber any other field that might be added to this doc later.
     */
    suspend fun recordAllTasksCleared(groupId: String): Int {
        val docRef = summaryDoc(groupId)
        return firestore.runTransaction { transaction ->
            val snapshot = transaction.get(docRef)
            val currentStreak = snapshot.getLong(FIELD_CURRENT_STREAK)?.toInt() ?: 0
            val lastDate = snapshot.getTimestamp(FIELD_LAST_COMPLETED_ALL_AT)?.toLocalDate()

            val today = LocalDate.now(ZoneId.systemDefault())
            if (lastDate == today) {
                return@runTransaction currentStreak
            }

            val yesterday = today.minusDays(1)
            val newStreak = if (lastDate == yesterday) currentStreak + 1 else 1

            transaction.set(
                docRef,
                mapOf(
                    FIELD_CURRENT_STREAK to newStreak,
                    FIELD_LAST_COMPLETED_ALL_AT to FieldValue.serverTimestamp(),
                ),
                SetOptions.merge(),
            )
            newStreak
        }.await()
    }

    private fun Timestamp.toLocalDate(): LocalDate =
        toDate().toInstant().atZone(ZoneId.systemDefault()).toLocalDate()
}
